### DISCUSSION TITLE
Putting my Solder where my Mouth Is

### DISCUSSION THREAD

**Discussion by mosqu-ito** [SpenceKonde/megaTinyCore#765] [Show and tell]
Recently I launched into a great spiel about using CCL for switch debouncing
and for frequency divide-by-two with 50/50 duty cycle.  I did this despite having
learned almost everything I know just this week.    :D
(Yes, I can be one of those.)

Anyway, it doesn't take long to put a switch and a couple of LEDs on a perf board.
So I did.  And, now I have an all-hardware debouncer and divider.  Unlike my last
little perf board, this demo is a sequential circuit (kinda the whole point).
And because it's sequential, to prove it works as advertised would require a short
video.  I'm not sure if I'm supposed to do that, even assuming I could produce such
a thing and figure out how to upload it.  So for now, I'm just going to give you a
still and ask you to trust me.

<pre>
//        DEBOUNCE and TOGGLE

#include "MyAliases3226.h"
#include "MyGeneral3226.h" 
#include "MyPins3226.h"
#include "MyChannels3226.h"
#include "MyLuts3226.h"

defU8 truthAlways = 0xFF;
defU8 truthNever  = 0x00;
defU8 truthRSal   = 0b11010100;

defU8 SwitchDown = ChipPin11;    // L2In0 alSet
defU8 SwitchUp   = ChipPin10;    // L2In1 alReset
defU8 BlueLED    = ChipPin08;    // L2Out DebounceOut
defU8 GreenLED   = ChipPin02;    // L0Out ToggleOut

void setup() {

   pinMode(SwitchDown,INPUT_PULLUP); pinMode(SwitchUp,INPUT_PULLUP);
   pinMode(GreenLED,  OUTPUT);       pinMode(BlueLED, OUTPUT);

   // Build RS Latch on Lut2...
   LutFeedback(2,2);                    // Input 2 is Q
   EnableLutPin(SwitchUp);              // Input 1 is R
   EnableLutPin(SwitchDown);            // Input 0 is S
   EnableLutPin(BlueLED);               // And lets monitor it.
   // I like to write all the truth bytes at the end when I start the Luts.

   // Build Toggler, a one bit counter with no carries      ;)
   ChannelFromLut(0,2); LutFromChannel(0,2, 0);     // Brings the clock over from the Debouncer.
   LocallyClockedJK(0);                             // I like to box things up.  But I'll open this one for you below.
   EnableLutPin(GreenLED);                          // And let's keep an eye on this too.
   
   StartLuts(truthNever,truthRSal,truthAlways,truthAlways);         // Lut3,Lut2,Lut1,Lut0

   }

void loop() {}
</pre>

And here's one of my little boxes...

I like to annotate where the information appeared in the datasheet.  (scroll far right)
Please don't trash the way I write to registers.  I'm old.  You'll hurt my feelings.  And then I'll die.

<pre>
void LocallyClockedJK(cU8 Lut) {
   if ((Lut==0)||(Lut==1))  {
      WriteRegField(adrCCL_LUT0CTRLA,1, 3,3);    // ClkSrc Input 2   "locally clocked"     datasheet 28.5.6
      WriteRegField(adrCCL_SEQCTRL0, 2, 3,4);    // Seq Type JK      "JK"                  datasheet 28.5.2
      }
   else {
      WriteRegField(adrCCL_LUT2CTRLA,1, 3,3);
      WriteRegField(adrCCL_SEQCTRL1, 2, 3,4);
      }
   }
</pre>

![DebTog](https://user-images.githubusercontent.com/106849308/184089359-30b17cf2-843b-41c3-a80c-11c8cab08ef5.jpg)

**Comment by SpenceKonde** [comment]
Don't put solder where your mouth is unless it's lead-free

**Comment by mosqu-ito** [comment]
Whenever I feel like I need a third hand, which is often, I rip off a foot of good old 60/40 and shove it right in there.  When I was little, my dad gave me half a pound of mercury for toy.

**Reply by SpenceKonde** [reply]
only a half pound? That aint much mercury, that stuff is dense. 

The real problem with mercury is twofold - when it gets into the environment it gets methylated by bacteria, and turns from mildly toxic to super toxic, and if left exposed to air, it evporates fast enough that it's a hazard. 
Cops a while back did a wellness check on this funeral director who hadn't been seen for a wile. Turns out he'd been stealing the amalgalm fillings from the bodies, which were mixtures of mercury gold and other noble metals. But no metals dealer is gonna touch that cause it''s obviously taken from a corpse. So he cooked off the mercury in his poorly ventilated basement. He was dead, and the home was terribly contaminated with mercury. It's  a shame, because it's so much more fun t play with than gallium.

**Comment by mosqu-ito** [comment]
I always suspected it had been sold by the full pound.  I'll bet he kept half of my mercury to play with himself.  He's lucky he's dead.

**Reply by mosqu-ito** [reply]
Uhm... accent on the second syllable of "himself"  ...I hope.

**Reply by SpenceKonde** [reply]
I think most likely it was sold as 250g, I've never seen mercury traded in imperial measures. 

Doubt you could have much fun with mercury in like _that_ even toxicity aside. Gallium or a GaIn/GaSn/GaInSn alloy would imo be a better bet. Notice I don't say a good bet - just better than mercury.

**Reply by mosqu-ito** [reply]
<pre>
        --
     /      \
  /            \
/       V        \
\     100mg      /
  \            /
     \      /
        --

* suitable for soft metals
</pre>
